webapp/src/app/projects/services/preferences.service.ts

```diff
@@ -7,13 +7,25 @@ export class PreferencesService {
   forProject = (projectId: String) => ({
```

Comment:
functions using `this` is a no no. Don't make me write reason, please.
You don't need `this` at all, anywhere here.
`.call(this, ...)` is a red flag

You need simple factory with a `projectId` closure.



```
forProject = (projectId: String) =>{
    return  ({
      referenceLanguages: {
        add,
        get,
        set,
        remove,
      },
    });

    function get() {
      const valueFromStorage = window.localStorage.getItem(`${projectId}.preferredReferenceLanguageCode`);
      let parsedValueFromStorage = [];
      try {
        parsedValueFromStorage = JSON.parse(valueFromStorage || '[]');
      } catch {
        console.log('Your reference locales setting has been reset')
      }

      return parsedValueFromStorage
    }

    function set(localeCodes){
        window.localStorage.setItem(`${projectId}.preferredReferenceLanguageCode`, JSON.stringify(localeCodes))
    }

    function add(localeCode: string) {
      const persistedLocales = get();
      if (persistedLocales.includes(localeCode)) {
        return;
      }

      set([...persistedLocales, localeCode]);
    }

    function remove(localeCode: string) {
      const persistedLocales = get();

      if (!persistedLocales.includes(localeCode)) {
        return;
      }

      const updatedPersistedLocales = persistedLocales.filter(persistedLocale => persistedLocale !== localeCode);

      if (updatedPersistedLocales.length > 0) {
        set(updatedPersistedLocales);
      } else {
        window.localStorage.removeItem(`${projectId}.preferredReferenceLanguageCode`);
      }
    }
  }
  ```